docs/testing/strategy.md

```diff
@@ -0,0 +1,187 @@
+
+```js
+const someFunc = (op, ...numbers) => {
+  let answer
+  switch (expr) {
+    case 'sum':
+      return numbers.reduce((a, b) => a + b, 0)
+    case 'avg':
+      return numbers.reduce((a, b) => a + b, 0) / numbers.length
+    case 'min':
+      return Math.min(...numbers)
+    case 'max':
+      return Math.max(...numbers)
+    default:
+      console.log('op not recognised')
+      return NaN
+  }
+}
+```
+
+This could be tested with a single test for each `op` parameter, this would suffice for code coverage but would not cover all of the applicable test cases. What happens if it was called with the following calls:
+
+```js
+   someFunc('avg', {numbers: [1,2,3]})  // Error caser
+   someFunc('tree', 1, 2, 3, 4) // Error case
+   someFunc('sum', [1, 2, 3, 4]) // Normal case
+   someFunc(false, 1, 2, 3, 4) // Error case
+   someFunc('max', 1, 2, 3, 4...) // 1 million int arguments, Extreme case
+```
+
+All five calls above would not be needed to test the function to a reasonable standard. We do not want to check every possible input as that would take far too long to both code and execute. Test cases can overlap and multiple cases can be executed from a single test. This can be written like so:
+
+```js
+describe('someFunc Tests', () => {
+  //Error
+  it('should handle incorrect op parameter', () => {
+    expect(someFunc('false', 1, 2, 3)).toBe(NaN)
+    expect(someFunc(false)).toThrow()
+  })
+
+  //Error
+  it('should handle incorrect numbers parameter', () => {
+    expect(someFunc('avg', [1, 2, 3])).toThrow()
+    expect(someFunc('avg')).toBe(NaN)
+  })
+
+  //Extreme
+  it('should be able to handle large amounts of input', () => {
+    expect(someFunc('sum' /* lots of numbers */)).toBe(/*correct result */)
+  })
+
+  //Normal
+  it('should return the average of a series of numbers passed in', () => {
+    expect(someFunc('avg', 1, 2, 3)).toBe(2)
+  }) // similar tests for normal cases for each op
+})
+```
+
+> **Note** The test cases type commented above each test
+
+### Mocking
+
+When we unit test a file / function / component everything not part of it should be mocked out. This is to prevent failures in other parts of the code mudying up our tests. For Exmple:
+
+```js
+import fileService from './fileService'
+
+const fileTitleNormaliser = fileId => {
+  const file = fileService.getFile(fileId)
+  file.title = file.title.toLowerCase()
+  return fileService.updateFile(file) // returns boolean true or false for success / failure
+}
+```
+
+In the above code the fileService may be third party so out of scope for our unit tests or it may ust be in another part of the repo, either way its a point of failure that we are not concerned with in this test so it needs to be mocked out:
+
+```js
+import fileService from './fileService'
```

Comment:
Should we mention it might be worth putting mocked services into a common repository of mocked things, so we can reuse it elsewhere... what are your thoughts on this?